### ISSUE TITLE
Question about data preprocessing

### ISSUE THREAD

**Issue by kakukakujirori** [microsoft/MoGe#44]
Hi, this is tangent from the main part, but I'd like to know how you batchfied the training images of different sizes.

The beginning of section 4 of the paper says:
> Our model is trained on images with aspect ratios uniformly sampled between 1: 2 and 2: 1 and pixel counts from 250K to 500K. We apply several image augmentations, including color jittering, Gaussian blurring, JPEG compression-decompression, and random cropping. The principal point of the cropped image is aligned with its center using a perspective crop.

I guess you first predefined the image size per batch and then adjusted the size of each image. However, because of the diversity in resolution and aspect ratio, it will result in a large area of zero padding or a too-distorted image aspect ratio (if you possibly used resizing), which might not be good for model training. I'm curious how you consistently managed all these different image sizes.

It will be okay to wait for the code release, but I would appreciate it if you could give even a short reply. Thank you.

**Comment by EasternJournalist** [comment]
Hi, Thanks for your interest! It is true that batched images must have the same size. We customized the dataset and dataloader so that image preprocessing (step 2) is done after batch sampling. This is the pipeline of our dataloader:

1. Sample a batch of images (of difference sizes), and determine a shared target size for this batch.
2. For each image, perform *perspective cropping* to get the cropped image of the same size. Note that the cropped image is not necessarily smaller than the original image, as the perspective cropping is done by `cv2.remap`. In this way, we can obtain arbitrary size of cropped image regardless of the size of original image.
3. Concatenate the cropped images as a batch for model input.

**Comment by kakukakujirori** [comment]
Thank you for your answer. I may not fully understand perspective cropping. 

To define a perspective transformation, we will first need to know at least 4 points of coordinates before and after the transformation. 

- Assuming the target 4 points are chosen from the corner points of the target image, are the source points randomly selected from the original image?
- The above operation surely doesn't introduce zero padding in the warped image, but the original aspect ratio can be severely distorted especially when the original image size and the target image size are very different. Did you tolerate it for domain generalization, or am I misunderstanding something?

Sorry to bother you. I'd appreciate your reply.

**Comment by EasternJournalist** [comment]
Hi. We have just released training code. You may find the details in [moge/train/dataloader.py](https://github.com/microsoft/MoGe/blob/main/moge/train/dataloader.py). I hope it would resolve your concern.

**Comment by kakukakujirori** [comment]
Thank you so much! There seems to be so much ingenuity involved, so I'll spend a few more days looking closely.
The question will be resolved by referring to the code, so I will close this issue. Thanks again for your great work!
